### PR TITLE
Add explanation for E0210 and revise E0117.

### PR DESCRIPTION
**Pull request by nham**
part of #24407 

I'm not sure whether I should be trying to explain the general rule in the E0210 explanation or just point people to the RFC. However, if we go with the latter option I think that the RFC will need to be revised slightly, since it is not quite as gentle as I would like.

Also, the link to RFC 1023 is not the correct one (it should be https://github.com/rust-lang/rfcs/blob/master/text/1023-rebalancing-coherence.md), but the correct one is too long. I'm aware of @michaelsproul's PR https://github.com/rust-lang/rust/pull/26290 from awhile back, but it doesn't seem to be working. Has there not been a new snapshot yet?

### CODE REVIEW

File: src/librustc_typeck/diagnostics.rs [hunk 1]
```diff
@@ -1833,6 +1847,71 @@ impl MyTrait for Foo {
...
+To see why that last example was allowed, you need to understand the general
+rule. Unfortunately this rule is a bit tricky to state. Consider an `impl`:
+
+```
+impl<P1, ..., Pm> ForeignTrait<T1, ..., Tn> for T0 { ... }
+```
+
+where `P1, ..., Pm` are the type parameters of the `impl` and `T0, ..., Tn`
+are types. One of the types `T0, ..., Tn` must be a local type (this is another
+orphan rule, see the explanation for E0117). Let `i` be the smallest integer
```
Comment (nham) [inline src/librustc_typeck/diagnostics.rs]: There's a spurious tidy failure due to the use of "E0117" in this line. I'm looking into how fixable it is

Comment (nham) [inline src/librustc_typeck/diagnostics.rs]: This is fixed now

**Comment by GuillaumeGomez** [comment]
No r+ here ?

**Comment by Gankro** [comment]
@bors r+ rollup

Thanks!

**Comment by GuillaumeGomez** [comment]
@nham: Can you update the spreadsheet (or once it'll be merged ? :p) please ?

**Comment by nham** [comment]
@GuillaumeGomez sure will.

**Pull request merged**